### ISSUE TITLE
`try_for_each` false positive

### ISSUE THREAD

**Issue by realbigsean** [rust-lang/rust-clippy#6599]
Lint name: map_collect_result_unit

I tried this code:

```rust
            logs.iter()
                .map(|raw_log| {
                    raw_log.to_deposit_log(self.inner.spec()).map_err(|error| {
                        Error::FailedToParseDepositLog {
                            block_range: block_range.clone(),
                            error,
                        }
                    })
                })
                .collect::<Result<Vec<_>, _>>()?
```

This iterator maps to a struct `DepositLog`, and the lint seems to be for mapping to `()`

It seems like `<Result<Vec<_>, _>>()` and `<Result<_, _>>()` get picked up as `<Result<(), _>()`

### Meta

- `cargo clippy -V`:  clippy 0.0.212 (e1884a8 2020-12-29)
- `rustc -Vv`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-unknown-linux-gnu
release: 1.49.0
```

**Comment by realbigsean** [comment]
nevermind, I was wrong. There was another iterator returning `<Result<(), _>()` later on
